test/compiled_metric_test.rb

```diff
@@ -0,0 +1,717 @@
+    assert_equal(2, shop_456_datagram.value)
+  end
+
+  def test_aggregates_static_tag_metrics
+    metric = StatsD::Instrument::CompiledMetric::Counter.define(
+      name: "foo.bar",
+      static_tags: { service: "web" },
+    )
+
+    metric.increment(value: 1)
+    metric.increment(value: 2)
+    metric.increment(value: 5)
+
+    @aggregator.flush
+
+    assert_equal(1, @sink.datagrams.size)
+    datagram = @sink.datagrams.first
+    assert_equal(8, datagram.value) # 1 + 2 + 5
+  end
+
+  def test_minimal_allocations_with_aggregation_static_tags
+    skip("Allocation tracking not supported on JRuby") if RUBY_ENGINE == "jruby"
+
+    metric = StatsD::Instrument::CompiledMetric::Counter.define(
+      name: "foo.bar",
+      static_tags: { service: "web" },
+    )
+
+    # Warm up to ensure everything is cached
+    metric.increment(value: 1)
+
+    # Measure allocations - disable GC to get accurate count
+    GC.disable
+    before = GC.stat(:total_allocated_objects)
+    metric.increment(value: 1)
+    after = GC.stat(:total_allocated_objects)
+    GC.enable
+
+    allocations = after - before
+    assert(allocations <= 2, "Expected <= 2 allocations with aggregation but got #{allocations}")
+  end
+
+  def test_minimal_allocations_with_aggregation_dynamic_tags
+    skip("Allocation tracking not supported on JRuby") if RUBY_ENGINE == "jruby"
+
+    metric = StatsD::Instrument::CompiledMetric::Counter.define(
+      name: "foo.bar",
+      tags: { shop_id: Integer },
+    )
+
+    # Warm up to ensure tag combination is cached
+    metric.increment(shop_id: 123, value: 1)
+
+    # Measure allocations on subsequent call with same tags - disable GC to get accurate count
+    GC.disable
+    before = GC.stat(:total_allocated_objects)
+    metric.increment(shop_id: 123, value: 1)
+    after = GC.stat(:total_allocated_objects)
+    GC.enable
+
+    allocations = after - before
+    assert(allocations <= 2, "Expected <= 2 allocations with aggregation (cached tags) but got #{allocations}")
+  end
+
+  def test_sample_rate_ignored_with_aggregation
+    # When aggregating, sample_rate should be ignored and the aggregated
+    # value should be emitted with sample_rate = 1.0 (or omitted entirely)
+    metric = StatsD::Instrument::CompiledMetric::Counter.define(
+      name: "foo.bar",
+      static_tags: { service: "web" },
+      sample_rate: 0.1,
+    )
+
+    # Even with sample_rate specified at definition time, aggregation ignores it
+    metric.increment(value: 5)
+    metric.increment(value: 3)
+
+    @aggregator.flush
+
+    assert_equal(1, @sink.datagrams.size)
+    datagram = @sink.datagrams.first
+    assert_equal("test.foo.bar", datagram.name)
+    assert_equal(8, datagram.value) # 5 + 3 (not scaled)
+    # Sample rate should be 1.0 (default) when aggregating
+    assert_equal(1.0, datagram.sample_rate)
+    refute_includes(datagram.source, "|@")
+  end
+end
+
+class DatagramBlueprintBuilderTest < Minitest::Test
```

Comment:
From what I can tell, all these tests have the `StatsD::Instrument::CompiledMetric` as the system under test. I think some can be reframed as `DatagramBlueprintBuilder.build`, others moved to classes above

Reply:
Moved them up 👍🏼